chain-network/src/data/p2p.rs

```diff
@@ -27,3 +30,77 @@ impl fmt::Display for Peer {
         write!(f, "{}", self.addr)
     }
 }
+
+const NODE_ID_LEN: usize = 32;
+
+#[derive(Copy, Clone, PartialEq, Eq, Hash)]
+pub struct NodeId([u8; NODE_ID_LEN]);
+
+impl fmt::Debug for NodeId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("NodeId(0x")?;
+        for byte in self.0.iter() {
+            write!(f, "{:02x}", byte)?;
+        }
+        f.write_str(")")
+    }
+}
+
+impl NodeId {
+    #[inline]
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.0
+    }
+
+    pub fn authenticated(self, signature: &[u8]) -> Result<AuthenticatedNodeId, Error> {
+        if signature.len() != NODE_SIGNATURE_LEN {
```

Comment:
The `VerificationAlgorithm` trait has `SIGNATURE_SIZE` within it. Can we reuse is somehow?

Reply:
That would bring a dependency on `chain-crypto` and all its dependencies into chain-network, as mentioned in https://github.com/input-output-hk/chain-libs/pull/445#issuecomment-698795009.

This, and also exporting verification API on `NodeId`, is perhaps better, to make the functionality self-contained.

Reply:
If so, then why not include it in the `Node` trait?

Reply:
That would entangle the `Node` trait into the signature of `NodeId`. Maybe `NodeId` should encapsulate a `Box<[u8]>` instead, and the same for the signature field, to keep things flexible and allow the application to choose an algorithm freely. It's not performance critical and the API uses allocated values elsewhere, anyway.